torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.cpp

```diff
@@ -0,0 +1,77 @@
+#include <caffe2/torch/testing/_internal/faulty_rpc_agent/faulty_process_group_agent.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+std::string fromVec(std::vector<char> vec) {
+  std::string s(vec.begin(), vec.end());
+  return s;
+}
+
+FaultyProcessGroupAgent::FaultyProcessGroupAgent(
+  std::string workerName,
+  std::shared_ptr<c10d::ProcessGroup> pg,
+  int numSendRecvThreads,
+  std::chrono::milliseconds rpcTimeout,
+  std::vector<std::string> messagesToFail,
+  int failNumSends)
+  : ProcessGroupAgent(workerName, pg, numSendRecvThreads, rpcTimeout),
+    failNumSends_(failNumSends) {
+  parseMessagesToFailInput(messagesToFail);
+}
+
+void FaultyProcessGroupAgent::parseMessagesToFailInput(std::vector<std::string> messagesToFail) {
+  // Since we can only pass strings corresponding to the Message Types from the
+  // python tests, we must parse the list of strings and resolve the actual
+  // types. We will then check this list of types in the send function to
+  // determine whether we should fail or not.
+  for (const auto& msgString : messagesToFail) {
+    if (msgString == "RREF_FORK_REQUEST") {
+      messageTypesToFail_.emplace(MessageType::RREF_FORK_REQUEST);
+    } else if (msgString == "RREF_CHILD_ACCEPT") {
+      messageTypesToFail_.emplace(MessageType::RREF_CHILD_ACCEPT);
+    } else if (msgString == "RREF_USER_DELETE") {
+      messageTypesToFail_.emplace(MessageType::RREF_USER_DELETE);
+    } else if (msgString == "CLEANUP_AUTOGRAD_CONTEXT_REQ") {
+      messageTypesToFail_.emplace(MessageType::CLEANUP_AUTOGRAD_CONTEXT_REQ);
+    }
+  }
+}
+
+std::shared_ptr<FutureMessage> FaultyProcessGroupAgent::send(const WorkerInfo& to, Message&& message) {
+  // We only fail control messages that have been specified by the test case.
+  // For all other messages, we just send them without any failures.
+  if (!shouldFailMessage(message.type())) {
+    return ProcessGroupAgent::send(to, std::move(message));
+  }
+  // This send function checks the failMessageCountMap_ to check whether
+  // we must fail the next send. If the send must be failed, we set an error
+  // on the returned future immediately and increment the counter in the map,
+  // otherwise we just call the ProcessGroupAgent send.
+  LOG(INFO) << "NUm fail sends " << failNumSends_;
```

Comment:
A bunch of this code is duplicated from the internal test that has the same check to fail via a map right? Could we possibly put that code somewhere generic and re-use it?

Reply:
This class is re-used in the internal ProcessGroupAgentTest - the changes are on Phabricator.